Set Yaw Action Server Architecture

aniket11bh opened the pull request:
- Action server completed for sending goal.
- Package `all_controllers` to include all types of controllers in future
- Removed clash of `controllerResult.msg` file.

TODO : 
- Shift `fuzzyControl.py` file to `manual_control` package.

aniket11bh committed:
Able to maintain the goal sent by the client.

- TODO : make it working for change in yaw by large angles, in -ve side.

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

icyflame commented on control_system_stack/setyaw/src/main.py:
```diff
@@ -0,0 +1,185 @@
+        pub_thrusters6.publish(thruster6Data)
+
+        # Debug messages
+        # rospy.logdebug("--------")
```
@aniket11bh You don't need to comment out `logdebug` lines. Uncomment these lines, so that if the node is run in `verbose` mode then we will be able to see these lines as well. Otherwise these lines won't show up.

icyflame commented on control_system_stack/setyaw/src/main.py:
```diff
@@ -0,0 +1,185 @@
+        FIRST_ITERATION = True
+        Current_yaw = -1
+
+        rospy.Subscriber(topicHeader.ABSOLUTE_RPY, absoluteRPY, self.imuCB)
```
@aniket11bh Move the subscriber to the `main` block of this file. Always declare all the subscribers and the publishers in the main block for easier access.

icyflame replied:
@aniket11bh I still think that all the subscribers and the publishers should be moved to the constructor (`__init__`) of this class. That way we will have faster access to the flow of data. (You can declare `pub_thrusters4` and `pub_thrusters6` as global variables and then, initialize them inside the constructor, along with this Subscriber.

icyflame commented on control_system_stack/setyaw/src/main.py:
```diff
@@ -0,0 +1,185 @@
+	pub_thrusters6 = rospy.Publisher(topicHeader.CONTROL_PID_THRUSTER6, thrusterData6Thruster, queue_size = 2)
+
+
+	rospy.init_node('setYaw')
```
@aniket11bh You need to do this as per the `verbose` code style. Check this [comment](https://github.com/auviitkgp/kraken_3.0/issues/70#issuecomment-144986277) for details on how to do this in Python.

Also, change the name of the node to `setyaw-action-server`.

icyflame replied:
@aniket11bh I have edited and added a [new comment](https://github.com/auviitkgp/kraken_3.0/issues/70#issuecomment-147985376), use this as the reference for all further packages.

icyflame commented on control_system_stack/setyaw/src/test_client.py:
```diff
@@ -0,0 +1,59 @@
+
+def serverSentSomeFeedback(feedback_obj):
+
+    rospy.loginfo('Recieved feedback from the server, Desired_yaw : %f', feedback_obj.Desired_yaw)
```
@aniket11bh I don't think that the all these values are required at every iteration. Just the error should do. The error is basically the difference between Desired and Current yaw anyway.

icyflame commented on control_system_stack/setyaw/src/main.py:
```diff
@@ -0,0 +1,185 @@
+
+
+	rospy.init_node('setYaw')
+	SetYaw(rospy.get_name())
```
@aniket11bh Write the name of this action server communication explicitly here. The value to be put here is `'setYaw'`. Having a variable such as this is very confusing.

aniket11bh committed:
Resolve issues for PR :

- Do `verbose-code-style` in main.py
- Action server communication named explicitly
- Change Action server node name to `setyaw-action-server`
- Do loginfo for Error as feedback only in client

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

aniket11bh committed:
Add Header header to action file of setYaw

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

icyflame commented:
Code works as expected. The graphing inside `rqt_plot` still does not work. @aniket11bh is finding solutions to this. We will be able to merge this PR after that change.

aniket11bh committed:
Add a publisher to plot the setYaw feedback data.

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

aniket11bh committed:
Change the declaration position of publisher and subscribers

- All the messages initialized inside the class
- All the pub and sub are inside the constructor

TODO : 
- Need to terminate the process of achieving a goal when, client decides.

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

aniket11bh committed:
Use the setYaw msg in fuzzycontrol.py for plotting.

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

icyflame commented:
LGTM

aniket11bh commented:
As one of the uses of `actionlib` package is that if the service takes a long time to execute, the user might want the ability to cancel the request during execution or get periodic feedback about how the request is progressing.
But it is not working right now, because these lines are commented out and the problem is that the there is some problem in working of the actionlib because of them : 
```python
if self._as.is_preempt_requested():
        rospy.loginfo('%s: Preempted' % self._action_name)
        self._as.set_preempted()
        success = False
        break
```

icyflame commented:
@aniket11bh So, why don't you uncommon to it and see what the problem is? I don't see why it is not working, maybe we are missing something.

icyflame commented on sensor_stack/seabotix/src/FuzzyControl/fuzzyParams.py:
```diff
@@ -0,0 +1,75 @@
+f_ssets = [[ # error
```
We can re-write this code later so as to use ROS' Parameter server to get these parameters.

The values are defined in a `yaml` file such as this one [config/dynamics_g500.yaml](https://github.com/uji-ros-pkg/underwater_simulation/blob/indigo-devel/underwater_vehicle_dynamics/config/dynamics_g500.yaml) and then accessed using a function in Python like [getParams](https://github.com/uji-ros-pkg/underwater_simulation/blob/indigo-devel/underwater_vehicle_dynamics/src/dynamics.py#L26)

aniket11bh committed:
Preempt done

Signed-off-by: Aniket Bhushan <aniket2701bhushan@gmail.com>

icyflame commented:
@aniket11bh 

Checked this on my machine, It works fine for most part. One of the problems that is still there is that on pressing `Ctrl+C` inside the client script, it says `You pressed Ctrl+C` but doesn't preempt the goal. This is not that big of an issue, so, I am going to go ahead and merge this PR in, but let's work on solving this as soon as the next dev cycle begins.

icyflame commented:
Fixed in https://github.com/auviitkgp/kraken_3.0/commit/051180f1a2d9bf8c8d510e739ab33223ebf47323
